File: source/3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */
#include "cpp_includes.h"
// @lc code=start
class Solution {
 public:
  int lengthOfLongestSubstring(string s) {
    int n = s.length();
    if (n < 2) return n;

    int left = 0;
    int right = 0;
    int maxlen = 0;

    unordered_set<char> charSet;
    while (right < n) {
      if (charSet.count(s[right]) == 0) {
        charSet.insert(s[right++]);
        maxlen = max(maxlen, right - left);
      } else {
        charSet.erase(s[left++]);
      }
    }
    return maxlen;
  }
};
// @lc code=end

File: source/3.无重复字符的最长子串.cs
/*
 * @lc app=leetcode.cn id=3 lang=csharp
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        if (s.Length < 2) return s.Length;

        var left = 0;
        var right = 0;
        var maxlen = 0;

        var charSet = new HashSet<char>();
        
        while (right < s.Length) {
            if (charSet.Contains(s[right]) == false) {
                charSet.Add(s[right++]);
                maxlen = Math.Max(maxlen, right - left);
            } else {
                charSet.Remove(s[left++]);
            }
        }

        return maxlen;
    }
}
// @lc code=end

